### ISSUE TITLE
Treat .stylelintrc as YAML or JSON

### ISSUE THREAD

**Issue by SevenOutman** [prettier/prettier#10920]
**Environments:**

- Prettier Version: 2.1.2
- Running Prettier via: CLI
- Runtime: Node.js 14.15.0
- Operating System: macOS
- Prettier plugins (if any): 

**Steps to reproduce:**

    $ prettier --write .eslintrc    
    .eslintrc 32ms
    $ prettier --write .stylelintrc 
    [error] No parser could be inferred for file: .stylelintrc

Prettier [treats `.eslintrc` (without extension) as JSON](https://github.com/prettier/prettier/blob/8f17187bf7b8aed10b923396f112b896a0a2023a/src/language-js/index.js#L94), but doesn't do the same to `.stylelintrc` or other cosmiconfig style config files.

**Expected behavior**

- Option 1: Add some more popular `*rc` as built-in supported.
- Option 2: Add a new `cosmiconfig` parser and use it as default for any `*rc` files without an extension.

**Comment by fisker** [comment]
Add it to https://github.com/prettier/prettier/blob/b84bc954e8725ad7e5afbd7a849bb9d9a507d274/src/language-yaml/index.js#L15 should work, PR welcome.

**Comment by SevenOutman** [comment]
> Add it to this list
> 
> https://github.com/prettier/prettier/blob/b84bc954e8725ad7e5afbd7a849bb9d9a507d274/src/language-yaml/index.js#L15
> 
> should work

I doubt if it's a good idea to add it to yaml list because a `.stylelintrc` can also be JSON format. 

I noticed that the yaml parser *can* handle JSON files but the results are not quite the same. Example:

```bash
$ prettier .eslintrc --parser json
{
  "extends": ["prettier"],
  "rules": {
    "react/display-name": "off",
    "react/prop-types": "off",
    "@typescript-eslint/camelcase": "off",
    "@typescript-eslint/no-empty-interface": "off"
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolver": {
      "webpack": {
        "config": "webpack.config.js"
      },
      "typescript": {}
    },
    "import/ignore": ["node_modules"]
  }
}
```

```bash
$ prettier .eslintrc --parser yaml
{
  'extends': ['prettier'],
  'rules':
    {
      'react/display-name': 'off',
      'react/prop-types': 'off',
      '@typescript-eslint/camelcase': 'off',
      '@typescript-eslint/no-empty-interface': 'off',
    },
  'settings':
    {
      'import/parsers': { '@typescript-eslint/parser': ['.ts', '.tsx'] },
      'import/resolver': { 'webpack': { 'config': 'webpack.config.js' }, 'typescript': {} },
      'import/ignore': ['node_modules'],
    },
}
```

**Comment by fisker** [comment]
Try rename this file to `.prettierrc`, you'll see the magic https://github.com/prettier/prettier/blob/b84bc954e8725ad7e5afbd7a849bb9d9a507d274/src/language-yaml/embed.js#L6

**Comment by SevenOutman** [comment]
> Try rename this file to `.prettierrc`, you'll see the magic

So after `.stylelintrc` is added to the yaml list, does it need be added in this ~~magic~~ hack condition too?

**Comment by fisker** [comment]
Yes, but only if confirmed it support both `yaml` and `json` syntax.

**Comment by SevenOutman** [comment]
> Yes, but only if confirmed it support both `yaml` and `json` syntax.

That's exactly what I suggested as a new `cosmiconfig` parser. 😂

**Comment by fisker** [comment]
I don't see why we need a new parser,

If it only supports `json`, add to `json` list.
If it only supports `yaml`, add to `yaml` list, don't to the trick.

**Comment by fisker** [comment]
I guess you mean, you can config `.unknownrc` to use `cosmiconfig`, but why don't just config to use `yaml` or `json`?

**Comment by SevenOutman** [comment]
> I guess you mean, you can config `.unknownrc` to use `cosmiconfig`, but why don't just config to use `yaml` or `json`?

What I suggest is Prettier adding built-in support for cosmiconfig style configs (indicated by `*rc` naming convention, which are treated as unknown files so far) so that users don't need to config `overrides` manually. 

A new `cosmiconfig` parser is just one possible implementation for this feature.

**Comment by fisker** [comment]
As far as I know, many `.xxrc` files are neither `yaml` nor `json`, we still need hard code those filenames.

**Comment by SevenOutman** [comment]
> As far as I know, many `.xxrc` files are neither `yaml` nor `json`, we still need hard code those filenames.

I see. Going to learn some magic now. :)

**Comment by fisker** [comment]
Fixed in #10924
